pkg/kubernetes/watcher.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+	"time"
+
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+
+	collector_metric "github.com/sustainable-computing-io/kepler/pkg/collector/metric"
+	"github.com/sustainable-computing-io/kepler/pkg/config"
+)
+
+const (
+	informerTimeout = time.Minute
+	podResourceType = "pods"
+)
+
+var (
+	regexReplaceContainerIDPrefix = regexp.MustCompile(`.*//`)
+)
+
+type ObjListWatcher struct {
+	// Lock to syncronize the collector update with the watcher
+	Mx *sync.Mutex
+
+	k8sCli       *kubernetes.Clientset
+	ResourceKind string
+	informer     cache.SharedInformer
+	stopChannel  chan struct{}
+
+	// ContainersMetrics holds all container energy and resource usage metrics
+	ContainersMetrics *map[string]*collector_metric.ContainerMetrics
+}
+
+func newK8sClient() *kubernetes.Clientset {
+	var restConf *rest.Config
+	var err error
+	if config.KubeConfig == "" {
+		// creates the in-cluster config
+		restConf, err = rest.InClusterConfig()
+		klog.Infoln("Using in cluster k8s config")
+	} else {
+		// use the current context in kubeconfig
+		restConf, err = clientcmd.BuildConfigFromFlags("", config.KubeConfig)
+		klog.Infoln("Using out cluster k8s config: ", config.KubeConfig)
+	}
+	if err != nil {
+		klog.Infoln("%v", err)
+		return nil
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(restConf)
+	if err != nil {
+		klog.Fatalf("%v", err)
+	}
+	return clientset
+}
+
+func NewObjListWatcher() *ObjListWatcher {
+	w := &ObjListWatcher{
+		stopChannel:  make(chan struct{}),
+		k8sCli:       newK8sClient(),
+		ResourceKind: podResourceType,
+	}
+	if w.k8sCli == nil {
+		return w
```

Comment:
should we fatal and error out here?

Reply:
No since we can run kepler without the watcher as the original code.
For example, for testing cases that kepler is not running in kubernetes.

Reply:
For e2e test case, we can export k8s config and run the test if possible https://github.com/sustainable-computing-io/kepler/blob/main/e2e/e2e_metric_test.go#L50 just for example. 
For unit test case, I am not sure ...